src/audio_unit/fft.rs

```diff
@@ -0,0 +1,45 @@
+use crate::{audio_unit::AudioUnit, util::number, Result};
+use rustfft::FftPlanner;
+
+pub struct Fft {
+    planner: FftPlanner<f32>,
+}
+
+impl Fft {
+    pub fn new() -> Self {
+        Self {
+            planner: FftPlanner::new(),
+        }
+    }
+}
+
+impl AudioUnit for Fft {
+    fn process(&mut self, input: &[f32], output: &mut [f32]) -> Result<()> {
+        let mut input_as_complex: Vec<_> = input.iter().copied().map(number::to_complex).collect();
+
+        let length = input.len();
+
+        self.planner
+            .plan_fft_forward(length)
+            .process(&mut input_as_complex);
```

Comment:
I'm guessing there's a "sweet spot" as far as how many samples you want to run an FFT against for it to give you high-fidelity results pretty efficiently? Do we know how that matches up with the typical size of an `input` chunk of samples? If `input` was not always an appropriate size, would that imply wanting to have eg a ring buffer that allowed you to run FFT's against something other than the most recent `input` chunk? I don't really grasp the basics of how it's doing the transformation back + forth (eg for the inverse how does it know how things changed over the course of the set of samples if all an FFT encodes is some "static" frequencies without anything indicating the time dimension)?

Reply:
> I'm guessing there's a "sweet spot" as far as how many samples you want to run an FFT against for it to give you high-fidelity results pretty efficiently? Do we know how that matches up with the typical size of an input chunk of samples? If input was not always an appropriate size, would that imply wanting to have eg a ring buffer that allowed you to run FFT's against something other than the most recent input chunk?

Yeah, at least based on my experiments in the past, the buffer size makes a pretty big difference. I think that ideally we want to have the audio library use our desired buffer size (I think that is within our control). Otherwise we'd need to add a frame of latency (unless our window size was always evenly divisible by our FFT size).

> I don't really grasp the basics of how it's doing the transformation back + forth (eg for the inverse how does it know how things changed over the course of the set of samples if all an FFT encodes is some "static" frequencies without anything indicating the time dimension)?

I think the key there is that it's not generating the sounds in the same way they were generated in real life. For example, if you have a buffer which has a sine wave at pitch A for the first half, and pitch B for the second half, running an FFT on that is going to give you a (large) list of frequencies/magnitudes, which, when played at once, result in something which sounds the same as the original input. It won't just give you back the frequency for A, and the frequency for B. (I'd guess that that means that the smaller your buffer size, the more likely your FFT will give you the frequencies that correspond to reality, since the buffer is less likely to contain frequency changes.)

This interactive article gives a good demo of how this is possible (but not really how it works): https://www.jezzamon.com/fourier/